src/base/static/utils/config-loader-utils.js

```diff
@@ -0,0 +1,53 @@
+// NOTE: These utility methods transform sections of the config for use by the app.
+// TODO: I don't like this very much, because it means sections of the config
+// aren't declarative. Ideally we should revisit this at some point and
+// refactor.
+
+// Transform the story section of the config to build the data structure we
+// need for story navigation.
+const transformStoryContent = storyConfig => {
```

Comment:
Related to my earlier comment, I think we should leverage this abstraction so that the implementation details of a Story's state are in one location. I would also consider renaming this function `storyFactory` or `hydrateStoryFromConfig`.

For example, I think it's confusing that we are creating multiple Story representations throughout our code. We have a representation returned by this function, then another representation saved in `this.stories` in our StorySidebar:

`this.stories = Object.keys(storyConfig).reduce((stories, storyName) => {`

then another representation in our StorySidebar:

```
      return {
        currentStoryHeader: storyConfig[foundStoryName].header,
        currentStoryDescription: storyConfig[foundStoryName].description,
        currentStoryName: foundStoryName,
        currentStoryChapters: foundStory,
        currentStoryRoute: joinedRoute,
        currentChapter: foundStory.get(joinedRoute),
      };
```

Instead, I think we should have a single representation of a Story that is hydrated from the config. And all of the logic to construct that representation should be in one place. Although I can imagine the only exception would be transforming the yml file to a json object - but that translation should be naive.

To help avoid mutations, you might want to explore returning an Immutable Record for this instead of a POJO: https://facebook.github.io/immutable-js/docs/#/Record

Reply:
Yes, this seems like an improvement. So I suppose `transformStoryContent` should be responsible then for the resolution to collection `Promise.all()` that currently lives in the story sidebar component. Any component wanting to use story content would have to wait on the resolution to those promises.

Reply:
final thought on this: I realize that we actually need the build time transformation in place, because it's used to set custom map state if we route to a place that's part of a story. This routing could happen before all the collection promises have resolved (i.e. if we route directly to a place in a story), so that information needs to be available immediately.

So for that reason, I think I'll keep the build time transformation in place, but move the creation of the story data structure to a separate utility and make it available to any component that might need it, not just the sidebar.